processing/src/main/java/io/druid/query/groupby/GroupByQueryQueryToolChest.java

```diff
@@ -162,47 +163,45 @@ public GroupByQueryQueryToolChest(
   )
   {
     // If there's a subquery, merge subquery results and then apply the aggregator
-
     final DataSource dataSource = query.getDataSource();
-
     if (dataSource instanceof QueryDataSource) {
       final GroupByQuery subquery;
       try {
-        // Inject outer query context keys into subquery if they don't already exist in the subquery context.
-        // Unlike withOverriddenContext's normal behavior, we want keys present in the subquery to win.
-        final Map<String, Object> subqueryContext = Maps.newTreeMap();
-        if (query.getContext() != null) {
-          for (Map.Entry<String, Object> entry : query.getContext().entrySet()) {
-            if (entry.getValue() != null) {
-              subqueryContext.put(entry.getKey(), entry.getValue());
-            }
-          }
-        }
-        if (((QueryDataSource) dataSource).getQuery().getContext() != null) {
-          subqueryContext.putAll(((QueryDataSource) dataSource).getQuery().getContext());
+        boolean pushDownQuery = shouldPushDownQuery(query);
+        if (pushDownQuery) {
+          GroupByQuery innerMostQuery = getInnerMostQuery(query);
+          GroupByQuery.Builder innerQueryBuilder = new GroupByQuery.Builder(innerMostQuery);
+          GroupByQuery.Builder pushDownQueryBuilder = new GroupByQuery.Builder(query);
+          // Unset the push down nested query flag so that the historical doesn't erroneously end up pushing down the query itself
+          pushDownQueryBuilder.overrideContext(ImmutableMap.of(GroupByQueryConfig.CTX_KEY_FORCE_PUSH_DOWN_NESTED_QUERY, false));
+          GroupByQuery queryToPushDown = pushDownQueryBuilder.build();
+          innerQueryBuilder.setQueryToPushDown(queryToPushDown);
```

Comment:
Would you elaborate more on this? `setQueryToPushDown()` means setting the outer query for the inner query?

Reply:
This was done to make sure that we can pass on the entire nested query over to the historical nodes. The role of the inner most query is only limited to figuring out what segments to query. But what gets executed on the historical nodes is this pushed down query which is nothing but the complete query. I think there is some improvement possible here. Right now, I am intercepting on DirectDruidClient to see whether the pushedDownQuery is present. If so, the DirectDruidClient gets hold of it and sends it over to the historical node. 

Relevant snippet in DirectDruidClient: https://github.com/druid-io/druid/pull/5471/files#diff-604c8c11338380d32a71a7621f031d2aR172